### ISSUE TITLE
[QUESTION] Type inference with decorators

### ISSUE THREAD

**Issue by zero323** [python/mypy#7659]
I have an use case that can simplified to something like this (that's `_typing.pyi` stub, but it doesn't seem to matter)

```python
from typing import Callable

ThatFunction = Callable[[], str]

f: Callable[[ThatFunction], Callable]      
```

where `f` is intended to be used as a decorator.

To my confusion if I apply it as follows Mypy (`0.740+dev.beec11a28fd835acb1e4ee8b2bf14d969c4e5922`)

```python
from _typing import f

@f       
def g():
    y: int = 1
    return y
```

doesn't seem to mind. 

Personally I'd expect something around:

```
error: Argument 1 has incompatible type "Callable[[], int]"; expected "Callable[[], str]"
Found 1 error in 1 file (checked 2 source files)
```

which is what I get if I explicitly annotate the return type

```python
from _typing import f

@f       
def g() -> int:
    return 1
```

Is it a bug or am I doing something off?

**Comment by gvanrossum** [comment]
Because you didn’t annotate g() in the first example, mypy assumes it returns Any, and then no error is produced. Check the mypy docs (very start).

**Comment by zero323** [comment]
Silly me. I am not sure why I was convinced that this should work. Thanks @gvanrossum.
